src/sparseml/optim/helpers.py

```diff
@@ -493,3 +494,83 @@ def _maybe_parse_number(val: str) -> Union[str, float, int]:
             return float(val)
         except Exception:
             return val
+
+
+def validate_metadata(metadata: dict, yaml_str: str) -> dict:
+    """
+    Compare the metadata carried over from the recipe (`yaml_str`) with
+    the new, incoming metadata ('metadata').
+
+    If yaml_str is a staged recipe, this function returns a multi-key
+        dict where:
+        -keys are stage names
+        -values are metadata for respective stage name
+    If yaml_str is a normal recipe, this function returns a single-key
+        dict where:
+        -key is '__metadata__'
+        -value is metadata for the recipe
+
+    If there exists a key in new metadata which also exists in previous metadata,
+    it will overwrite the old metadata. However, it is prohibitive to pass novel keys.
+
+    :param metadata: New metadata
+    :param yaml_str: String representation of the recipe YAML file,
+        (may contains previous metadata)
+    :return: Validated metadata
+    """
+
+    default_metadata_key = "__metadata__"
+    container = load_recipe_yaml_str_no_classes(yaml_str)
+    is_recipe_staged = check_if_staged_recipe(container)
+
+    if is_recipe_staged:
+        stage_names = [
+            stage_name
+            for stage_name, stage_dict in container.items()
+            if isinstance(stage_dict, dict)
```

Comment:
this technically isn't sufficient for if an item is a stage or not, we also need to check if stage_dict has a key with "modifiers" in its name

Reply:
Good catch, corrected.